File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTEventSubscribeService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimflutter.ResultCallback
import com.netease.nimflutter.SafeResult
import com.netease.nimflutter.convertToEvent
import com.netease.nimflutter.toMap
import com.netease.nimlib.sdk.NIMClient
import com.netease.nimlib.sdk.Observer
import com.netease.nimlib.sdk.RequestCallbackWrapper
import com.netease.nimlib.sdk.ResponseCode
import com.netease.nimlib.sdk.event.EventSubscribeService
import com.netease.nimlib.sdk.event.EventSubscribeServiceObserver
import com.netease.nimlib.sdk.event.model.Event
import com.netease.nimlib.sdk.event.model.EventSubscribeRequest
import com.netease.nimlib.sdk.event.model.EventSubscribeResult

class FLTEventSubscribeService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {

    private val onEventObserver =
        Observer { eventList: List<Event> ->
            run {
                notifyEvent(
                    "observeEventChanged",
                    mutableMapOf("eventList" to eventList.map { it.toMap() }.toList())
                )
            }
        }

    init {
        nimCore.onInitialized {
            NIMClient.getService(EventSubscribeServiceObserver::class.java)
                .observeEventChanged(onEventObserver, true)
        }
    }

    override val serviceName = "EventSubscribeService"

    override fun onMethodCalled(method: String, arguments: Map<String, *>, safeResult: SafeResult) {
        when (method) {
            "registerEventSubscribe" -> registerEventSubscribe(
                arguments,
                ResultCallback(safeResult)
            )
            "unregisterEventSubscribe" -> unregisterEventSubscribe(
                arguments,
                ResultCallback(safeResult)
            )
            "batchUnSubscribeEvent" -> batchUnSubscribeEvent(arguments, ResultCallback(safeResult))
            "publishEvent" -> publishEvent(arguments, ResultCallback(safeResult))
            "querySubscribeEvent" -> querySubscribeEvent(arguments, ResultCallback(safeResult))
            else -> safeResult.notImplemented()
        }
    }

    private fun registerEventSubscribe(
        arguments: Map<String, *>,
        resultCallback: ResultCallback<List<String>>
    ) {
        var eventSubscribeRequest = convertToEventSubscribeRequest(arguments)
        if (eventSubscribeRequest.eventType > 0) {
            NIMClient.getService(EventSubscribeService::class.java)
                .subscribeEvent(eventSubscribeRequest).setCallback(
                    object : RequestCallbackWrapper<List<String>>() {
                        override fun onResult(
                            code: Int,
                            result: List<String>?,
                            exception: Throwable?
                        ) {
                            // result代表部分订阅失败的用户id
                            resultCallback.result(
                                NimResult(code = 0, data = result) {
                                    mutableMapOf("resultList" to it)
                                }
                            )
                        }
                    }
                )
        } else {
            resultCallback.result(
                NimResult(
                    code = -1,
                    errorDetails = "eventType must be greater than 0"
                )
            )
        }
    }

    private fun unregisterEventSubscribe(
        arguments: Map<String, *>,
        resultCallback: ResultCallback<Void>
    ) {
        var eventSubscribeRequest = convertToEventSubscribeRequest(arguments)
        if (eventSubscribeRequest.eventType > 0) {
            NIMClient.getService(EventSubscribeService::class.java)
                .unSubscribeEvent(eventSubscribeRequest)
            resultCallback.result(NimResult(code = 0))
        } else {
            resultCallback.result(
                NimResult(
                    code = -1,
                    errorDetails = "eventType must be greater than 0"
                )
            )
        }
    }

    private fun batchUnSubscribeEvent(
        arguments: Map<String, *>,
        resultCallback: ResultCallback<Void>
    ) {
        var eventSubscribeRequest = convertToEventSubscribeRequest(arguments)
        if (eventSubscribeRequest.eventType > 0) {
            NIMClient.getService(EventSubscribeService::class.java)
                .batchUnSubscribeEvent(eventSubscribeRequest)
            resultCallback.result(NimResult(code = 0))
        } else {
            resultCallback.result(
                NimResult(
                    code = -1,
                    errorDetails = "eventType must be greater than 0"
                )
            )
        }
    }

    private fun publishEvent(arguments: Map<String, *>, resultCallback: ResultCallback<Void>) {
        var event = convertToEvent(arguments)
        if (event.eventType > 0) {
            NIMClient.getService(EventSubscribeService::class.java).publishEvent(event)
            resultCallback.result(NimResult(code = 0))
        } else {
            resultCallback.result(
                NimResult(
                    code = -1,
                    errorDetails = "eventType must be greater than 0"
                )
            )
        }
    }

    private fun querySubscribeEvent(
        arguments: Map<String, *>,
        resultCallback: ResultCallback<List<EventSubscribeResult>>
    ) {
        var eventSubscribeRequest = convertToEventSubscribeRequest(arguments)
        if (eventSubscribeRequest.eventType > 0) {
            NIMClient.getService(EventSubscribeService::class.java)
                .querySubscribeEvent(eventSubscribeRequest).setCallback(
                    object : RequestCallbackWrapper<List<EventSubscribeResult>>() {
                        override fun onResult(
                            code: Int,
                            result: List<EventSubscribeResult>?,
                            exception: Throwable?
                        ) {
                            // result代表部分订阅失败的用户id
                            if (code == ResponseCode.RES_SUCCESS.toInt()) {
                                resultCallback.result(
                                    NimResult(
                                        0,
                                        data = result,
                                        convert = { list ->
                                            mutableMapOf(
                                                "eventSubscribeResultList" to list.map { it.toMap() }
                                                    .toList()
                                            )
                                        }
                                    )
                                )
                            } else {
                                resultCallback.result(
                                    NimResult(
                                        code = -1,
                                        errorDetails = "query error"
                                    )
                                )
                            }
                        }
                    }
                )
        }
    }

    private fun convertToEventSubscribeRequest(arguments: Map<String, *>): EventSubscribeRequest {
        var request = EventSubscribeRequest()
        if (arguments != null) {
            request.eventType = (arguments["eventType"] as Number).toInt()
            request.expiry = (arguments.getOrElse("expiry") { 0L } as Number).toLong()
            request.isSyncCurrentValue =
                arguments.getOrElse("syncCurrentValue") { false } as Boolean
            request.publishers = arguments["publishers"] as List<String>
        }
        return request
    }
}

File: nim_core_v2/ios/Classes/services/FLTEventSubscribeService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NIMSDK

enum EventSubscribeType: String {
  case RegisterEventSubscribe = "registerEventSubscribe"
  case UnregisterEventSubscribe = "unregisterEventSubscribe"
  case BatchUnSubscribeEvent = "batchUnSubscribeEvent"
  case PublishEvent = "publishEvent"
  case QuerySubscribeEvent = "querySubscribeEvent"
}

class FLTEventSubscribeService: FLTBaseService, FLTService {
//    override init() {
//        super.init()
//        NIMSDK.shared().subscribeManager.add(self)
//    }

  override func onInitialized() {
    NIMSDK.shared().subscribeManager.add(self)
  }

  deinit {
    NIMSDK.shared().subscribeManager.remove(self)
  }

  // MARK: Service Protocol

  func serviceName() -> String {
    ServiceType.EventSubscribeService.rawValue
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any],
                      _ resultCallback: ResultCallback) {
    switch method {
    case EventSubscribeType.RegisterEventSubscribe.rawValue:
      registerEventSubscribe(arguments, resultCallback)
    case EventSubscribeType.UnregisterEventSubscribe.rawValue:
      unregisterEventSubscribe(arguments, resultCallback)
    case EventSubscribeType.PublishEvent.rawValue:
      publishEvent(arguments, resultCallback)
    case EventSubscribeType.QuerySubscribeEvent.rawValue:
      querySubscribeEvent(arguments, resultCallback)
    case EventSubscribeType.BatchUnSubscribeEvent.rawValue:
      batchUnSubscribeEvent(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  // MARK: Public Method

  func registerEventSubscribe(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let request = NIMSubscribeRequest.yx_model(with: arguments) else {
      let result = NimResult(nil, -1, "registerEventSubscribe but the args are invalid")
      resultCallback.result(result.toDic())
      return
    }
    if request.type < 0 {
      let result = NimResult(
        nil,
        -1,
        "registerEventSubscribe eventType must be greater than 0"
      )
      resultCallback.result(result.toDic())
      return
    }
    NIMSDK.shared().subscribeManager.subscribeEvent(request) { error, ret in
      guard let nserror = error as NSError? else {
        if ret != nil, ret!.count > 0 {
          let result = NimResult(["resultList": ret], 0, nil)
          resultCallback.result(result.toDic())
        } else {
          let result = NimResult(nil, 0, nil)
          resultCallback.result(result.toDic())
        }
        return
      }
      let result = NimResult(nil, NSNumber(value: nserror.code), nserror.description)
      resultCallback.result(result.toDic())
    }
  }

  func unregisterEventSubscribe(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let request = NIMSubscribeRequest.yx_model(with: arguments) else {
      let result = NimResult(nil, -1, "unregisterEventSubscribe but the args are invalid")
      resultCallback.result(result.toDic())
      return
    }
    if request.type < 0 {
      let result = NimResult(
        nil,
        -1,
        "unregisterEventSubscribe eventType must be greater than 0"
      )
      resultCallback.result(result.toDic())
      return
    }
    NIMSDK.shared().subscribeManager.unSubscribeEvent(request) { error, ret in
      guard let nserror = error as NSError? else {
        if ret != nil, ret!.count > 0 {
          let result = NimResult(ret, 0, nil)
          resultCallback.result(result.toDic())
        } else {
          let result = NimResult(nil, 0, nil)
          resultCallback.result(result.toDic())
        }
        return
      }
      let result = NimResult(nil, NSNumber(value: nserror.code), nserror.description)
      resultCallback.result(result.toDic())
    }
  }

  func publishEvent(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let event = NIMSubscribeEvent.createSubscribeEvent(arguments) else {
      let result = NimResult(nil, -1, "publishEvent but the args are invalid")
      resultCallback.result(result.toDic())
      return
    }
    if event.type < 0 {
      let result = NimResult(nil, -1, "publishEvent eventType must be greater than 0")
      resultCallback.result(result.toDic())
      return
    }

    NIMSDK.shared().subscribeManager.publishEvent(event, completion: { error, ret in
      guard let nserror = error as NSError? else {
        resultCallback.result(NimResult.success().toDic())
        return
      }
      let result = NimResult(nil, NSNumber(value: nserror.code), nserror.description)
      resultCallback.result(result.toDic())
    })
  }

  func querySubscribeEvent(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let request = NIMSubscribeRequest.yx_model(with: arguments) else {
      let result = NimResult(nil, -1, "querySubscribeEvent but the args are invalid")
      resultCallback.result(result.toDic())
      return
    }
    if request.type < 0 {
      let result = NimResult(nil, -1, "querySubscribeEvent eventType must be greater than 0")
      resultCallback.result(result.toDic())
      return
    }
    NIMSDK.shared().subscribeManager.querySubscribeEvent(request) { error, ret in
      guard let nserror = error as NSError? else {
        if ret == nil {
          let result = NimResult(nil, 0, nil)
          resultCallback.result(result.toDic())
        } else {
          // 可以用map，后续优化
          var array: [[String: Any?]] = .init()
          for subscribeResult in ret! {
            if let s = subscribeResult as? NIMSubscribeResult,
               let dic = s.toDic() {
              array.append(dic)
            }
          }
          let result = NimResult(["eventSubscribeResultList": array], 0, nil)
          resultCallback.result(result.toDic())
        }
        return
      }
      let result = NimResult(nil, NSNumber(value: nserror.code), nserror.description)
      resultCallback.result(result.toDic())
    }
  }

  func batchUnSubscribeEvent(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let request = NIMSubscribeRequest.yx_model(with: arguments) else {
      let result = NimResult(nil, -1, "batchUnSubscribeEvent but the args are invalid")
      resultCallback.result(result.toDic())
      return
    }
    if request.type < 0 {
      let result = NimResult(
        nil,
        -1,
        "batchUnSubscribeEvent eventType must be greater than 0"
      )
      resultCallback.result(result.toDic())
      return
    }
    // publisher不传则为取消所有
    request.publishers = [Any]()
    NIMSDK.shared().subscribeManager.unSubscribeEvent(request) { error, ret in
      guard let nserror = error as NSError? else {
        if ret != nil, ret!.count > 0 {
          let result = NimResult(ret, 0, nil)
          resultCallback.result(result.toDic())
        } else {
          let result = NimResult(nil, 0, nil)
          resultCallback.result(result.toDic())
        }
        return
      }
      let result = NimResult(nil, NSNumber(value: nserror.code), nserror.description)
      resultCallback.result(result.toDic())
    }
  }
}

extension FLTEventSubscribeService: NIMEventSubscribeManagerDelegate {
  func onRecvSubscribeEvents(_ events: [Any]) {
    // 可以用map，后续优化
    var array = [[String: Any?]]()
    for evnet in events {
      if let e = evnet as? NIMSubscribeEvent,
         let dic = e.toDic() {
        array.append(dic)
      }
    }
    notifyEvent(serviceName(), "observeEventChanged", ["eventList": array])
  }
}
